### DISCUSSION TITLE
Using Zenoh from webassembly

### DISCUSSION THREAD

**Discussion by SverreEplov** [eclipse-zenoh/roadmap#69] [Q&A] [answered]
Hi,

I'm working on a project where all data are communicated from data gathering devices to a central data server using zenoh. 

There will be a webassembly based website developed in Rust using egui.rs for frontend, and I am wondering if it is possible to use zenoh to communicate data to a webassembly program running in a browser. 

Has anyone attempted to test/use zenoh in such a setup ? 

Best Regards,
 /Sverre Eplov

**Comment by gabrik** [comment, marked as answer]
Hi, @SverreEplov we indeed started experimenting with WASM. 
For the time being, we started with [Zenoh-Pico](https://github.com/eclipse-zenoh/zenoh-pico), for which there is a PR [#135](https://github.com/eclipse-zenoh/zenoh-pico/pull/135) that enables to compile to WASM with the support of WebSocket as a transport.

We are also investigating how to build the WASM from the Rust implementation. It first requires a review of external dependencies, as some of the ones we are using currently do not compile for the WASM target.

Is your use-case explicitly require a Zenoh rust-based client to compile as WASM (I guess so as you are using egui.rs) or a JS/WASM API will be able to fit?

**Reply by SverreEplov** [reply]
Hi @gabrik ,

I don't think a JS/WASM solution will cut it, as there will be at least 2 video streams plus a number of data/control topics to handle. It is encouraging to hear that a WASM version of the native Rust implementation is in the works. 

For the usecase I have, a client functionality will suffice, as the browser solution will work in client mode only. 

The part with reviewing/solving external dependencies for WASM use sounds very familiar - that is an issue for us as well, as we want to do a lot in the browser. (Video decoding, stream synchronization and so on). 

My plan right now is to find an intermediate solution, and switch to pure Zenoh when a Rust WASM solution is in the works. 

If this can be of any value, we will be happy to work with testing a WASM Zenoh in setup replicating a final production system during development of Zenoh in WASM.

Best Regards,
 /Sverre Eplov

**Comment by gabrik** [comment]
Just to keep track of the current status:
- https://github.com/eclipse-zenoh/zenoh/pull/419 Allows `no_std` for Zenoh (rust), which is one step towards enabling Zenoh to build for `wasm32-wasi` target. 
- https://github.com/eclipse-zenoh/zenoh/pull/424
